nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+		}
+	}
+	e.Status = stResolved
+	e.Flags = flags
+	e.putTime = now
+	e.validUntil = now.Add(time.Second * time.Duration(minTtl))
+	e.reply = *reply
+
+	if shouldNotify {
+		close(e.waitChan) // notify all the waiters by closing the channel
+	}
+}
+
+// wait until a valid reply is set in the cache
+func (e *cacheEntry) waitReply(request *dns.Msg, timeout time.Duration, now time.Time) (*dns.Msg, error) {
+	if e.Status == stResolved {
+		return e.getReply(request, now)
+	}
+
+	if timeout > 0 {
+		select {
+		case <-e.waitChan:
+			return e.getReply(request, now)
+		case <-time.After(time.Second * timeout):
+			return nil, errTimeout
+		}
+	}
+
+	return nil, errCouldNotResolve
+}
+
+func (e *cacheEntry) close() {
+	if e.Status == stPending {
+		close(e.waitChan)
+	}
+}
+
+// entriesSlice is used for sorting entries
+type cacheEntriesSlice []*cacheEntry
+
+func (p cacheEntriesSlice) Len() int           { return len(p) }
+func (p cacheEntriesSlice) Less(i, j int) bool { return p[i].validUntil.Before(p[j].validUntil) }
+func (p cacheEntriesSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+type cacheKey struct {
+	question dns.Question
+	protocol dnsProtocol
+}
+
+type entries map[cacheKey]*cacheEntry
+
+// Cache is a thread-safe fixed capacity LRU cache.
+type Cache struct {
+	Capacity int
+
+	entries entries
+	lock    sync.RWMutex
+}
+
+// NewCache creates a cache of the given capacity
+func NewCache(capacity int) (*Cache, error) {
+	if capacity <= 0 {
+		return nil, errInvalidCapacity
+	}
+	c := &Cache{
+		Capacity: capacity,
+		entries:  make(entries, capacity),
+	}
+	return c, nil
+}
+
+// Clear removes all the entries in the cache
+func (c *Cache) Clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entries = make(entries, c.Capacity)
+}
+
+// Purge removes the old elements in the cache
+func (c *Cache) Purge(now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.removeOldest(c.Capacity, now)
+}
+
+// Add adds a reply to the cache.
+func (c *Cache) Put(request *dns.Msg, reply *dns.Msg, proto dnsProtocol, flags uint8, now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if ent, found := c.entries[key]; found {
+		ent.setReply(reply, flags, now)
+	} else {
+		// If we will add a new item and the capacity has been exceeded, make some room...
+		if len(c.entries) >= c.Capacity {
+			c.removeOldest(1, now)
+		}
+		c.entries[key] = newCacheEntry(&question, reply, stResolved, now)
+	}
+}
+
+// Look up for a question's reply from the cache.
+// If no reply is stored in the cache, it returns a `nil` reply and no error. The caller can then `Wait()`
+// for another goroutine `Put`ing a reply in the cache.
+func (c *Cache) Get(request *dns.Msg, proto dnsProtocol, now time.Time) (reply *dns.Msg, err error) {
```

Comment:
> I think our DNS client to the fallback server in WeaveDNS is using a user-configurable buffer size (UDPSize) and I think this is used for announcing the maximum payload size with EDNS

It is: https://github.com/zettio/weave/pull/361. It's set to 4KB.

A little wiresharking suggests that libresolv won't use (or advertise) a larger buffer by default. So it's certainly a possibility that we get a non-truncated UDP response from upstream which we can't send downstream without truncating; and that we could answer a subsequent TCP request with the answer we got via UDP (if we didn't segregate answers according to how we obtained them).

> But I don't think we should assume clients support EDNS

Aye, we can only send larger responses if the client has itself advertised a larger buffer. Using EDNS0 is an option in resolv.conf, and "security-aware resolvers" per DNSSEC are required to use EDNS0 OPT RRs and larger buffer sizes (RFC 4035). I don't know how common those are.

> I don't think we should use these EDNS extensions for other parts of the system either.

That ought to be logged as an issue.

> what happens if we send requests to ns1.google.com announcing, for example, a buffer size of 65535 bytes?

WeaveDNS doesn't do that; why would it?


In sum: the current algorithm is

 * For a UDP request:
   * Do we have a cached UDP response? If so send that, truncating if necessary
      * If not, ask upstream, cache the result, and send it downstream, truncating if necessary
   * For a TCP request:
    * Do we have a cached TCP response? If so send that
      * If not, ask upstream over TCP, cache the result and send it downstream

Without segregating cache entries, the algorithm would be

 * For a request:
    * Do we have a cached response? If so, send that, truncating if necessary
      * If not, ask upstream*, cache the result, and send it downstream, truncating if necessary

*There's two tactics wrt using TCP upstream: it could be done in response to the client using TCP, or it could be immediately on getting a truncated result.